### PR TITLE
V03/smoothemulatorscott

### PR DESCRIPTION
**Pull request by wildsm**
This is the general pull request to bring in @scottedwardpratt 's SmoothEmulator package.

It will start as a draft PR and then reviews will be done here. Please assign yourself as a reviewer if you would like to sign up to test it.

### CODE REVIEW

**Comment by wildsm** [comment]
Marking as ready for review and seeking reviewers.

**Comment by wildsm** [comment]
@DanielRPhillips for socialization?

**Comment by scottedwardpratt** [comment]
Hi, The problem is that the utilities are not particularly related to the emulator project. I use them across many projects.

I planned on updating the utilities in the band repo by simply doing occasional copies of source and include files from my “real” utilities repo.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Scott Pratt                                        ***@***.*** %
% FRIB                                               +1-517-908-7460 %
% Michigan State University                            Office   2133 %
% East Lansing, MI 48824    http://nscl.msu.edu/directory/pratt.html %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%




















On Oct 3, 2023, at 11:01 AM, Stefan M Wild ***@***.***> wrote:



@scottedwardpratt<https://urldefense.com/v3/__https://github.com/scottedwardpratt__;!!HXCxUKc!1dUQTKgBjZW12fRvWeRngFaTRR42tWgYG50-1tuR_Vi7YSbfcjsV6je5oj5Y8foaJD8Ccbx9zB93oHcZxHzNHHg$> - is there any way that we can make SmoothUtilities part of SmoothEmulator?
If you want to keep them fully separate, then I suggest having part of the install procedure for SmoothEmulator to be to clone the external dependency (probably within SmoothEmulator).

**Comment by wildsm** [comment]
@scottedwardpratt - what is the location of the repo containing smoothutilities? 

I am proposing that the user be instructed to git clone this repo from GITHOME_BAND/SmoothEmulator (so that it would show up as a subdirectory of GITHOME_BAND/SmoothEmulator). In general, it would be good for everything needed to be brought in for smoothemulator to be in GITHOME_BAND/SmoothEmulator and not in a parallel file location.

**Comment by wildsm** [comment]
I am proceeding assuming that the instruction will reflect cloning and building commonutils (see below for my suggested way to proceed).

Here are the steps I took:

```
# Environmental variable
export GITHOME_BAND="/home/wild/oldcomp/repos/bandframework/software"
cd $GITHOME_BAND/SmoothEmulator

# Confirm cmake is around
which cmake

# Download and build eigen 
git clone https://gitlab.com/libeigen/eigen.git
cd eigen
mkdir build_dir
cd build_dir
cmake ../
sudo make install
cd ../..

# Download and build GSL 
sudo apt install libgsl-dev
export GSL_INCLUDE_DIR="/usr/include/gsl/"

# Obtain and build utilities
cd $GITHOME_BAND/SmoothEmulator
git clone git@github.com:scottedwardpratt/commonutils.git
cd commonutils/software
cmake .
make
```

`make` is failing due to the error

```
[  5%] Building CXX object CMakeFiles/msu_commonutils.dir/src/Arrays/arraycalc.cc.o
In file included from /home/wild/oldcomp/repos/bandframework/software/SmoothEmulator/commonutils/software/include/msu_commonutils/arrays.h:4,
                 from /home/wild/oldcomp/repos/bandframework/software/SmoothEmulator/commonutils/software/src/Arrays/arraycalc.cc:1:
/home/wild/oldcomp/repos/bandframework/software/SmoothEmulator/commonutils/software/include/msu_commonutils/commondefs.h:27:10: fatal error: Eigen/Dense: No such file or directory
   27 | #include <Eigen/Dense>
      |          ^~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/msu_commonutils.dir/build.make:76: CMakeFiles/msu_commonutils.dir/src/Arrays/arraycalc.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/msu_commonutils.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
```
I believe that this is because the correct Eigen3 include is not being used.
Suggestions for proceeding @scottedwardpratt ?

**Comment by scottedwardpratt** [comment]
Stefan, What platform are you on?



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Scott Pratt                                        ***@***.*** %
% FRIB                                               +1-517-908-7460 %
% Michigan State University                            Office   2133 %
% East Lansing, MI 48824    http://nscl.msu.edu/directory/pratt.html %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%




















On Oct 3, 2023, at 4:22 PM, Stefan M Wild ***@***.***> wrote:



I am proceeding assuming that the instruction will reflect cloning and building commonutils (see below for my suggested way to proceed).

Here are the steps I took:

# Environmental variable
export GITHOME_BAND="/home/wild/oldcomp/repos/bandframework/software"
cd $GITHOME_BAND/SmoothEmulator

# Confirm cmake is around
which cmake

# Download and build eigen
git clone https://gitlab.com/libeigen/eigen.git
cd eigen
mkdir build_dir
cd build_dir
cmake ../
sudo make install
cd ../..

# Download and build GSL
sudo apt install libgsl-dev
export GSL_INCLUDE_DIR="/usr/include/gsl/"

# Obtain and build utilities
cd $GITHOME_BAND/SmoothEmulator
git clone ***@***.***:scottedwardpratt/commonutils.git
cd commonutils/software
cmake .
make


make is failing due to the error

[  5%] Building CXX object CMakeFiles/msu_commonutils.dir/src/Arrays/arraycalc.cc.o
In file included from /home/wild/oldcomp/repos/bandframework/software/SmoothEmulator/commonutils/software/include/msu_commonutils/arrays.h:4,
                 from /home/wild/oldcomp/repos/bandframework/software/SmoothEmulator/commonutils/software/src/Arrays/arraycalc.cc:1:
/home/wild/oldcomp/repos/bandframework/software/SmoothEmulator/commonutils/software/include/msu_commonutils/commondefs.h:27:10: fatal error: Eigen/Dense: No such file or directory
   27 | #include <Eigen/Dense>
      |          ^~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/msu_commonutils.dir/build.make:76: CMakeFiles/msu_commonutils.dir/src/Arrays/arraycalc.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/msu_commonutils.dir/all] Error 2
make: *** [Makefile:91: all] Error 2


I believe that this is because the correct Eigen3 include is not being used.
Suggestions for proceeding @scottedwardpratt<https://urldefense.com/v3/__https://github.com/scottedwardpratt__;!!HXCxUKc!0O_ARjznKMGFE547TLCEO9H3OSIx5t3Ms6dKpFzZ-5p7lWZGpS_qskGnFwgKWufjwouc5xzq4J00Y8rHxlnTXBg$> ?

**Comment by wildsm** [comment]
I'm testing on Linux (Ubuntu), bash shell

**Comment by scottedwardpratt** [comment]
Hi Stefan,  I tried everything on my Linux. I did fix a CMakeLists.txt error, but that was for compiling SmoothEmulator, not SmoothUtilities.

It looks like it is some kind of Eigen problem, maybe with the version. I don’t think it’s a problem with Cmake, but that’s possible too. My undergrad might have had a similar problem. I’ll try figuring it out with him. Eigen is just supposed to be a bunch of header files, so I find this confusing.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Scott Pratt                                        ***@***.*** %
% FRIB                                               +1-517-908-7460 %
% Michigan State University                            Office   2133 %
% East Lansing, MI 48824    http://nscl.msu.edu/directory/pratt.html %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%




















On Oct 3, 2023, at 8:45 PM, Stefan M Wild ***@***.***> wrote:



I'm testing on Linux (Ubuntu), bash shell

**Comment by wildsm** [comment]
Is there a way to override/augment the Eigen include path in the utilities cmake?

If you see in the process I followed, 
export GSL_INCLUDE_DIR="/usr/include/gsl/"

helped ensure that the correct GSL path was there (this is something I think might need to go into the manual), but an
export EIGEN3_INCLUDE_DIR="/usr/local/include/eigen3/"
did not solve my problem... maybe because EIGEN3_INCLUDE_DIR is being overridden by the cmake(list)?

**Comment by scottedwardpratt** [comment]
I read some stuff and looked at the way eigen was installed on my Mac and on my Ubuntu.

It appears that Eigen installations tend to jump around over time. Sometimes /usr/include/eigen3/Eigen, or /usr/include/eigen, or …

I think this has confused CMake as people have some dead directories from former eigen installations.

I’ll try editing the CMakeLists.txt file to echoic for the environmental variable you mentioned, and if it’s set it will use that value. This will be a little hard to test though, because it was working on my installations.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Scott Pratt                                        ***@***.*** %
% FRIB                                               +1-517-908-7460 %
% Michigan State University                            Office   2133 %
% East Lansing, MI 48824    http://nscl.msu.edu/directory/pratt.html %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%




















On Oct 3, 2023, at 9:52 PM, Stefan M Wild ***@***.***> wrote:



Is there a way to override/augment the Eigen include path in the utilities cmake?

If you see in the process I followed,
export GSL_INCLUDE_DIR="/usr/include/gsl/"

helped ensure that the correct GSL path was there (this is something I think might need to go into the manual), but an
export EIGEN3_INCLUDE_DIR="/usr/local/include/eigen3/"
did not solve my problem... maybe because EIGEN3_INCLUDE_DIR is being overridden by the cmake(list)?

**Comment by scottedwardpratt** [comment]
Hi,  I think I provided fix.

I edited CMakeLists.txt, changing include statement to

include_directories(SYSTEM ${GSL_INCLUDE_DIR} $ENV{GSL_INCLUDE_DIR} ${EIGEN3_INCLUDE_DIR} $ENV{EIGEN3_INCLUDE_DIR} include)

I believe/hope that by adding the $ENV{……} arguments that this will add the paths represented by those environmental variables should the CMake find package screw up.

These should be pushed now. I’ll change the documentation to tell the user they should set these environmental variables should CMake fail.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Scott Pratt                                        ***@***.*** %
% FRIB                                               +1-517-908-7460 %
% Michigan State University                            Office   2133 %
% East Lansing, MI 48824    http://nscl.msu.edu/directory/pratt.html %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%




















On Oct 3, 2023, at 9:52 PM, Stefan M Wild ***@***.***> wrote:



Is there a way to override/augment the Eigen include path in the utilities cmake?

If you see in the process I followed,
export GSL_INCLUDE_DIR="/usr/include/gsl/"

helped ensure that the correct GSL path was there (this is something I think might need to go into the manual), but an
export EIGEN3_INCLUDE_DIR="/usr/local/include/eigen3/"
did not solve my problem... maybe because EIGEN3_INCLUDE_DIR is being overridden by the cmake(list)?

**Comment by wildsm** [comment]
Great, I'll try to work on this tomorrow/this week.

While you're updating the pdf, here are some initial thoughts:

- I would pull in commonutils via a git clone as done in the above; doing this in a subdirectory actually works well.
- Looking at commonutils, it looks like you are also assuming that Boost is installed; I would add that to the dependencies. 
- The pdf's title is PHY 321 Lecture Notes...
- typo "homeebrew"

**Comment by wildsm** [comment]
For me it did not fix the eigen include issue. So I made an edit to the CMakeLists.txt so that if a user has set an environment variable it will also include that.

Proceeding further with build and test, I run into a similar issue building SmoothEmulator, but by correcting environment variable (MSU -> BAND), it worked. This has been pushed.

The CMakeLists.txt in the templates was a doozy, I made significant updates to facilitate use (e.g., it was assuming the program was contained in "smooth"); a user now specifies an environment variable BUILD_DIR for where they should build.

So finally we get to my current error while running make in MY_LOCAL/build:

```
home/wild/oldcomp/repos/bandframework/software/SmoothEmulator/MY_LOCAL/main_programs/pca_readinfo_calcy_main.cc: In function ‘int main(int, char**)’:
/home/wild/oldcomp/repos/bandframework/software/SmoothEmulator/MY_LOCAL/main_programs/pca_readinfo_calcy_main.cc:20:8: error: ‘class PCA’ has no member named ‘ReadPCATransformationInfo’; did you mean ‘ReadTransformationInfo’?
   20 |   pca->ReadPCATransformationInfo();
      |        ^~~~~~~~~~~~~~~~~~~~~~~~~
      |        ReadTransformationInfo
make[2]: *** [CMakeFiles/pca_readinfo_calcy.dir/build.make:79: CMakeFiles/pca_readinfo_calcy.dir/home/wild/oldcomp/repos/bandframework/software/SmoothEmulator/MY_LOCAL/main_programs/pca_readinfo_calcy_main.cc.o] Error 1
make[2]: Leaving directory '/home/wild/oldcomp/repos/bandframework/software/SmoothEmulator/MY_LOCAL/build'
make[1]: *** [CMakeFiles/Makefile2:230: CMakeFiles/pca_readinfo_calcy.dir/all] Error 2
make[1]: Leaving directory '/home/wild/oldcomp/repos/bandframework/software/SmoothEmulator/MY_LOCAL/build'
make: *** [Makefile:94: all] Error 2
```

If I follow the prompt (i.e., delete "PCA"), it builds! I have pushed this but @scottedwardpratt should confirm this is OK.

Proceeding.

**Comment by wildsm** [comment]
@scottedwardpratt I attempted to run the simplex sampler (Section 8.4), but it resulted in a segmentation fault. Now stuck. 

Below I am including my full workflow for the next intrepid soul.

```
# Environmental variable
export GITHOME_BAND="/home/wild/oldcomp/repos/bandframework/software"
cd $GITHOME_BAND/SmoothEmulator

# Confirm cmake is around
which cmake

# Download and build eigen 
git clone https://gitlab.com/libeigen/eigen.git
cd eigen
mkdir build_dir
cd build_dir
cmake ../
sudo make install
cd ../..

# Download and build GSL 
sudo apt install libgsl-dev
export GSL_INCLUDE_DIR="/usr/include/gsl/"

# Build utilities; vital for my system was to include this environment variable
cd $GITHOME_BAND/SmoothEmulator/SmoothUtilities/software
export CUSTOM_EIGEN3_INCLUDE_DIR="/usr/local/include/eigen3/"
cmake .
make

# Build SmoothEmulator
cd $GITHOME_BAND/SmoothEmulator/software
cmake .
make

# Set up MY_PROJECT and MY_LOCAL
cd $GITHOME_BAND/SmoothEmulator
cp -r $GITHOME_BAND/SmoothEmulator/templates/myproject MY_PROJECT
cp -r $GITHOME_BAND/SmoothEmulator/templates/mylocal MY_LOCAL

# Build in MY_LOCAL; vital is to set this environment variable!
export BUILD_DIR="/home/wild/oldcomp/repos/bandframework/software/SmoothEmulator/MY_LOCAL"
cd $GITHOME_BAND/SmoothEmulator/MY_LOCAL/build
cmake .
make simplex
make smoothy_tune
make smoothy_calcobs

# Run Simplex Sampler
cd $GITHOME_BAND/SmoothEmulator/MY_PROJECT
../MY_LOCAL/bin/simplex parameters/simplex_parameters.txt
```

**Comment by wildsm** [comment]
@scottedwardpratt - we're going to postpone this for v0.3.0 and can include in v0.3.1 as soon as ready.

File: software/SmoothEmulator/software/CMakeLists.txt [hunk 1]
```diff
@@ -12,7 +12,9 @@ find_package(Eigen3 REQUIRED)
 
 message("eigen3_include_dir is ${EIGEN3_INCLUDE_DIR}")
 
-include_directories(SYSTEM include ${GITHOME_MSU}/SmoothEmulator/SmoothUtilities/software/include ${GSL_INCLUDE_DIR} $ENV{GSL_INCLUDE_DIR} ${EIGEN3_INCLUDE_DIR} $ENV{EIGEN3_INCLUDE_DIR} include)
+include_directories(SYSTEM include ${GITHOME_BAND}/SmoothEmulator/software/include ${GITHOME_BAND}/SmoothEmulator/SmoothUtilities/software/include ${GSL_INCLUDE_DIR} $ENV{GSL_INCLUDE_DIR} ${EIGEN3_INCLUDE_DIR} $ENV{EIGEN3_INCLUDE_DIR} include)
+
+message("${GITHOME_MSU}/SmoothEmulator/SmoothUtilities/software/include")
```
Comment (wildsm) [inline software/SmoothEmulator/software/CMakeLists.txt:17]: GITHOME_MSU should be GITHOME_BAND

**Comment by wildsm** [comment]
@scottedwardpratt I brought this branch up to date with develop and made some changes to the readmes to ensure spellcheck was successful.

**Comment by wildsm** [comment]
Scott says:

it’s ready to be looked at.

To clone the repo:
% git clone https://github.com/bandframework/bandframework.git -b v03/smoothemulatorscott

If you go to the doc/ directory you can look at the User Manual, which has installation instructions, plus a tutorial chapter.

Thanks to Stefan and Kyle for previous checks.

Please let me know what problems you might be having.

All the best — Scott
it’s ready to be looked at.

To clone the repo:
% git clone https://github.com/bandframework/bandframework.git -b v03/smoothemulatorscott

If you go to the doc/ directory you can look at the User Manual, which has installation instructions, plus a tutorial chapter.

Thanks to Stefan and Kyle for previous checks.

Please let me know what problems you might be having.

All the best — Scott

**Comment by wildsm** [comment]
@kylegodbey - are you reviewing this one?

**Comment by kylegodbey** [comment]
> @kylegodbey - are you reviewing this one?

I will, once the yada yada's subside

**Comment by scottedwardpratt** [comment]
Hi, It’s all ready. The only things that we changed had to do with pybind11, which is not mentioned in this release, and does not interfere with anything else.

Thanks!

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Scott Pratt                                        ***@***.*** %
% FRIB                                               +1-517-908-7460 %
% Michigan State University                            Office   2133 %
% East Lansing, MI 48824    http://nscl.msu.edu/directory/pratt.html %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%




















On Sep 13, 2024, at 10:53 AM, Kyle Godbey ***@***.***> wrote:



@kylegodbey<https://urldefense.com/v3/__https://github.com/kylegodbey__;!!HXCxUKc!wa_RWc9Vw1OLijvgFs6Z3h08tijAjMz3D6IM9SjFm_eB3ciZ53z4dFMrKaega9g2LRaZju1UeUW46dgtNbPDpBA$> - are you reviewing this one?

I will, once the yada yada's subside

**Comment by kylegodbey** [comment]
A couple notes on the manual and process:
- Maybe listing the C++ standard required somewhere would be a good idea -- I was building on an older Ubuntu version and had to fast forward to g++-11
- The `cmake -D YOUR_PREFERRED_COMPILER`  instruction led to an error for me, so I instead did `-D CMAKE_CXX_COMPILER=g++-11`
- It is claimed that `fakerhic` makes a directory `${MY PROJECT}/fullmodel_testdata/`, but for me it was in `smooth_data`. That may be user error.
- For the smooth emulator parameter file, perhaps put the name of the file in the manual's tutorial. Right now it says: *For the template used in this tutorial, that file is*
- Starting in section 10.2, there are some LaTeX errors -- perhaps due to `\end{verbatime}`
- Manual says `figs/YvY/` but it should be `figs/YvsY/`
- My `fullmodel_testdata` directory discrepancy shows up again for the YvsY business. Perhaps I went awry 
- The tutorial sections are a bit messed up. Looking in the ToC, we see section 8, then 9 and 10 that should be part of the tutorial, before jumping back to 9.

Other than that, I was able to run everything fine with no bugs. I think it's good to merge once some of the tutorial instructions are polished up.

**Comment by scottedwardpratt** [comment]
THANKS Kyle!

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Scott Pratt                                        ***@***.*** %
% FRIB                                               +1-517-908-7460 %
% Michigan State University                            Office   2133 %
% East Lansing, MI 48824    http://nscl.msu.edu/directory/pratt.html %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%




















On Sep 18, 2024, at 5:17 PM, Kyle Godbey ***@***.***> wrote:



A couple notes on the manual and process:

  *   Maybe listing the C++ standard required somewhere would be a good idea -- I was building on an older Ubuntu version and had to fast forward to g++-11
  *   The cmake -D YOUR_PREFERRED_COMPILER instruction led to an error for me, so I instead did -D CMAKE_CXX_COMPILER=g++-11
  *   It is claimed that fakerhic makes a directory ${MY PROJECT}/fullmodel_testdata/, but for me it was in smooth_data. That may be user error.
  *   For the smooth emulator parameter file, perhaps put the name of the file in the manual's tutorial. Right now it says: For the template used in this tutorial, that file is
  *   Starting in section 10.2, there are some LaTeX errors -- perhaps due to \end{verbatime}
  *   Manual says figs/YvY/ but it should be figs/YvsY/
  *   My fullmodel_testdata directory discrepancy shows up again for the YvsY business. Perhaps I went awry
  *   The tutorial sections are a bit messed up. Looking in the ToC, we see section 8, then 9 and 10 that should be part of the tutorial, before jumping back to 9.

Other than that, I was able to run everything fine with no bugs. I think it's good to merge once some of the tutorial instructions are polished up.

**Review by kylegodbey — APPROVED** [review]

**Comment by wildsm** [comment]
Thanks @kylegodbey & @scottedwardpratt

**Pull request merged**